[Compose 페이먼츠 3단계] 미플 미션 제출합니다.

HamBeomJoon opened the pull request:
# 셀프 체크리스트
- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?
- 2단계 마지막에 리뷰받은 supportingText, 문자열 분리, modifier위치 변경 등도 적용해서 변경 line이 많아보이네요 😅
- BankType에 Type마다 값들을 미리 설정?해놓는 크루들도 보이는데 해당 방법이 좋은건지 궁금합니다. 
- 일단 UiState, StateHolder에 대해 좀 더 공부해봐야할거같아서 2단계때 남긴 것들은 반영못한게 있는데, 스킵부탁드립니다.

## 스크린샷

https://github.com/user-attachments/assets/4d5f7749-d5a2-45f5-8f7e-99ebd6eaa834

HamBeomJoon committed:
docs: 3단계 기능 요구 사항 정의

HamBeomJoon committed:
feat: 에러 메시지를 AddCardError enum 클래스에서 관리

HamBeomJoon committed:
refactor: InputField 파라미터 순서 변경

HamBeomJoon committed:
refactor: TextField error 메시지 supportingText로 변경

HamBeomJoon committed:
refactor: uiModel 내부로 포매팅로직 이동

HamBeomJoon committed:
refactor: AddCardStateHolder 및 AddCardUiState 초기값 및 타입 변경

- AddCardStateHolder의 initialState 기본값 설정
- AddCardUiState의 cardNumber, expired, password 타입을 non-nullable로 변경하고 초기값 설정
- 관련 로직에서 nullable 타입 처리 제거

HamBeomJoon committed:
refactor: AddCardBox private로 변경

HamBeomJoon committed:
feat: BankType enum class 추가

HamBeomJoon committed:
feat: 카드사별 이미지 추가

HamBeomJoon committed:
feat: 카드사별 Enum Class에 displayName 추가

HamBeomJoon committed:
feat: 카드사 선택 BottomSheet 화면 추가

HamBeomJoon committed:
refactor: BankType enum에서 displayName 속성 제거

`BankType` enum 클래스에서 각 은행 타입의 표시 이름을 나타내는 `displayName` 속성을 제거했습니다.

HamBeomJoon committed:
feat: 카드사, 배경색 표시 기능 추가

카드사 정보를 `CardCompanyUiModel`로 분리하고, 카드 배경색을 표시하는 기능을 추가했습니다.

HamBeomJoon committed:
feat: BankType을 CardCompanyUiModel로 변환하는 mapper 추가

HamBeomJoon committed:
feat: 카드사별 색상 정의

HamBeomJoon committed:
feat: AddCardScreen에 은행 선택 Bottom Sheet 추가

HamBeomJoon committed:
feat: addCardUiState에 카드사UiModel 추가

HamBeomJoon committed:
feat: 카드 배경색 및 카드 정보 미리보기 추가

- 카드 회사에 따른 카드 배경색을 적용합니다.
- 우리 카드와 카카오뱅크 카드의 미리보기를 추가합니다.

HamBeomJoon committed:
feat: addCard StateHolder에 카드사 정보 저장 및 복원

HamBeomJoon committed:
feat: 은행 아이콘 클릭 시 카드 추가 화면으로 이동하는 기능 구현

HamBeomJoon committed:
feat: 카드에 카드사 이름 추가 및 UI 개선

- 카드 전면에 카드사 이름을 표시하도록 추가
- GOLD 색상을 Color.kt에 정의하여 사용
- UI 레이아웃 및 정렬 방식 개선

HamBeomJoon committed:
refactor: 은행 선택 BottomSheet 화면을 열었을 때 보이도록 수정

HamBeomJoon committed:
refactor: 은행 선택 바텀 시트 은행 목록 로직 변경

- 기존에 정의된 `bankList`를 사용하는 대신 `BankType.entries`를 사용하도록 변경
- `BankSelectRow` 컴포저블 함수를 `BankSelectBottomSheet` 내부로 통합
- 프리뷰 함수 이름 및 파라미터 변경

HamBeomJoon committed:
feat: 카드사 로고 이미지 추가 및 카드사 선택 BottomSheet 개선

- CardCompanyUiModel에 로고 리소스 ID 필드 추가
- BankSelectBottomSheet에서 카드사 선택 시 바로 BottomSheet가 닫히도록 로직 수정
- BankSelectBottomSheet에서 카드사 선택 상태를 내부에서 관리하지 않도록 변경
- 카드사 선택 BottomSheet가 초기에 항상 보이도록 수정
- 카드 정보 저장 및 복원 시 카드사 로고 정보도 포함하도록 BundleKeys 및 AddCardStateHolder 수정
- Mapper 함수에 각 카드사별 로고 리소스 ID 매핑 추가

HamBeomJoon committed:
test: AddCardScreenTest, CardListScreenTest 테스트에서 사용할 fixture 구현

HamBeomJoon committed:
test: BankItem, BankSelectBottomSheet 테스트 코드 추가

HamBeomJoon committed:
refactor: CardCompany -> Bank로 이름 변경

HamBeomJoon committed:
docs: 기능 요구 사항 완료 체크

HamBeomJoon committed:
refactor: CardListScreen 레이아웃 최적화

- CardListContent의 content 파라미터를 ColumnScope의 확장 함수로 변경하여 Spacer 추가를 간결하게 처리했습니다.
- 카드 목록의 스크롤 가능 여부를 카드의 개수에 따라 동적으로 결정하도록 수정했습니다.
- 불필요한 Spacer를 제거하고 카드 목록 하단에 일관된 간격을 추가했습니다.

HamBeomJoon committed:
refactor: 카드 리스트 UiState에서 enableScroll 제거

카드 리스트 UiState에서 `enableScroll` 프로퍼티를 제거합니다. `showAddButton` 프로퍼티는 기존 `get()` 방식에서 초기화 시 값을 할당하도록 변경합니다.

HamBeomJoon committed:
feat: 카드 목록이 없을 때, EmptyCardListContent 뷰를 표시하도록 수정

HamBeomJoon committed:
refactor: Mapper.kt 파일 삭제 및 toPresentation() 확장 함수 위치 변경

- `Mapper.kt` 파일을 삭제합니다.
- `BankType.toBankUiModel()` 확장 함수를 `CardUiModel.kt`로 이동하고 `BankType.toPresentation()`으로 이름을 변경합니다.
- `PaymentCard` 컴포저블의 padding 값을 수정합니다.
- `BankSelectBottomSheet` 컴포저블의 `dragHandle` 속성을 추가합니다.

HamBeomJoon committed:
chore: 기타 파일 정리

HamBeomJoon committed:
refactor: sheetState AddCardScreen으로 올리기

HamBeomJoon committed:
fix: 카드 선택 전까지 sheet 못 내리게 방지

HamBeomJoon committed:
refactor: cardListUiState에 showAddButton 계산로직 stateHolder로 이동

HamBeomJoon committed:
refactor: 카드 추가 화면 상태 저장 및 복원 로직 개선

- `AddCardStateHolder`에서 UI 상태를 `Bundle`에 저장하고 복원하는 로직을 개선했습니다.
    - 기존에는 각 필드를 개별적으로 저장하고 복원했지만, 이제 `BankUiModel`과 `AddCardError` 목록을 `Parcelable`로 직접 저장하고 복원합니다.
    - `AddCardError` enum에 `@Parcelize` 어노테이션을 추가하고 `Parcelable`을 구현했습니다.
    - `BundleKeys` 대신 직접 문자열 키를 사용하도록 변경했습니다.
- 입력 필드 컴포저블(`CardNumberInputField`, `ExpiredInputField`, `CardOwnerInputField`, `PasswordInputField`)에서 도메인 모델 대신 문자열을 직접 사용하도록 변경했습니다.
    - 이에 따라 `onValueChange` 콜백의 파라미터 타입도 변경되었습니다.
- `AddCardUiState`에서 유효성 검사 로직을 `AddCardStateHolder`로 이동하고, `submitted` 상태를 추가하여 제출 시에만 오류를 표시하도록 변경했습니다.
- `IntentsExtension.kt`에 `Bundle`에서 `Parcelable` 및 `ParcelableArrayList`를 안전하게 가져오기 위한 확장 함수를 추가했습니다.

HamBeomJoon committed:
refactor: AddCardStateHolder의 Bundle key 상수화 및 카드 유효성 검사 로직 개선

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/BankItem.kt:
```diff
@@ -0,0 +1,60 @@
+    ) {
+        Image(
+            painter = painterResource(id = bank.logoRes),
+            contentDescription = stringResource(R.string.bank_item_content_description),
```
`은행 아이템 이미지`라고 고정되어 있으면 8개의 은행을 구분할 수 없을 것 같아요.
`%s`를 사용해서 은행이 구분되도록 하면 좋을 것 같아요.

HamBeomJoon replied:
8f5f0260c5660d15bbd22ffc37a8a110093f2ba4

변경하였습니다

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/BankItem.kt:
```diff
@@ -0,0 +1,60 @@
+import woowacourse.payments.ui.model.toPresentation
+
+@Composable
+fun BankItem(
```
`modifier`가 없네요 😮

HamBeomJoon replied:
헉 그러네요..
d5d1c757d93aed035d0fcca8bb693dbb72fd88dd

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/BankSelectBottomSheet.kt:
```diff
@@ -0,0 +1,64 @@
+    onDismiss: () -> Unit,
+) {
+    ModalBottomSheet(
+        onDismissRequest = { onDismiss() },
```
`onDismissRequest`는 언제 호출될까요?

HamBeomJoon replied:
onDismissRequest는 바텀 시트를 숨기기 위해 사용자나 시스템이 요청할 때 실행됩니다.

- 외부 영역 탭: 바텀 시트 바깥 영역을 탭했을 때.
- 아래로 스와이프: 바텀 시트를 아래로 드래그해서 완전히 내렸을 때.
- 하드웨어 뒤로 가기 버튼: 안드로이드 기기의 하드웨어 뒤로 가기 버튼을 눌렀을 때.

이러한 요청이 발생하면 ModalBottomSheet 컴포넌트가 onDismissRequest에 지정된 람다 함수를 실행합니다. 일반적으로 이 람다 함수는 바텀 시트의 가시성을 제어하는 상태 변수를 false로 변경하는 역할을 합니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/BankSelectBottomSheet.kt:
```diff
@@ -0,0 +1,64 @@
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun BankSelectBottomSheet(
```
현재 `sheetState.hide()`를 사용하지 않고 있어서 
아이템을 클릭 하면 한순간에 바텀 시트가 사라지도록 되어있어요.

`hide()`를 호출한 뒤 `showBottomSheetState`를 `false`로 변경하면 자연스러운 UI가 될 수 있을 것 같아요.

HamBeomJoon replied:
오 좋은 UX 감사합니다.
c7cb2156efbeda934850fe08cd54f055a6269095

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/CardOwnerInputField.kt:
```diff
@@ -1,87 +1,85 @@
+        supportingText = {
+            Row(
+                modifier = Modifier.fillMaxWidth(),
+                horizontalArrangement = Arrangement.SpaceBetween,
+            ) {
                 Text(
-                    text = stringResource(R.string.card_owner_placeholder),
-                    color = Color.LightGray,
+                    text = if (error != null) stringResource(error.messageRes) else "",
+                    modifier =
+                        Modifier.semantics {
+                            contentDescription =
+                                context.getString(R.string.card_owner_error_content_description)
+                        },
+                    color = Color.Red,
                 )
```
Text에 색상을 주는 방법도 있지만 `OutlineTextField`의 매개변수에 아래와 같은 값도 존재합니다.
```kotlin
 colors: TextFieldColors = OutlinedTextFieldDefaults.colors()
``` 
그래서 따로 Text에 색상을 정의하지 않아도  `OutlinedTextField`의 기본 색상으로 적용됩니다.

HamBeomJoon replied:
9447ac9065fbbbf6bc41de0b57e1d6aaff44e2a5
기본 색상을 적용했습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardStateHolder.kt:
```diff
@@ -10,64 +10,99 @@ import woowacourse.payments.domain.CardNumber
+                                    ?: emptySet(),
                         )
                     AddCardStateHolder(restoredState)
                 },
```
CardUiModel에 Parcelable를 적용하면 아래와 같이 간단하게 처리가 가능합니다.

```kotlin
val saver: Saver<AddCardStateHolder, CardUiModel> =
    Saver(
        save = { holder -> holder.uiState },
        restore = { state -> AddCardStateHolder(state) }
    )
```

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListStateHolder.kt:
```diff
@@ -7,20 +7,30 @@ import androidx.compose.runtime.setValue
 
     companion object {
         val saver =
             Saver<CardListStateHolder, List<CardUiModel>>(
                 save = { holder -> holder.uiState.cards },
-                restore = { cards -> CardListStateHolder(CardListUiState(cards)) },
+                restore = { cards -> CardListStateHolder() },
             )
```
`saver`가 정상 작동하지 않을 것 같아요 😮

HamBeomJoon replied:
헉 그러네요..

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListStateHolder.kt:
```diff
@@ -7,20 +7,30 @@ import androidx.compose.runtime.setValue
 
 class CardListStateHolder(
-    initialState: CardListUiState = CardListUiState(),
+    initialCards: List<CardUiModel> = emptyList(),
```
`AddCardStateHolder`에서는 `uiState`를 매개변수로 받았는데
여기서는 `initialCards`를 받아 내부에서 `uiState`를 생성하고 있네요
어떤 차이가 있나요?

HamBeomJoon replied:
AddCardStateHolder는 여러 독립적인 상태들(카드 번호, 만료일, 소유자 이름, 그리고 비밀번호)을 한 번에 관리해야 합니다.  또, 유효성 검사 에러 상태나 사용자가 선택한 은행 정보도 함께 관리해야 합니다.

CardListStateHolder의 상태는 본질적으로 하나의 핵심 데이터, 즉 카드 목록입니다. showAddButton과 같은 다른 상태는 이 카드 목록의 크기에 따라 결정되는 파생된 상태입니다. 따라서 CardListStateHolder는 List<CardUiModel>만 가지고 UI가 필요할 때마다 목록의 크기를 확인하여 showAddButton 값을 계산하면 되기 때문입니다

moondev03 replied:
말씀해주신 `CardListStateHolder`에 대한 설명은 잘못된 내용인 것 같습니다.
코드를 살펴보면 `CardListStateHolder`는 `CardListUiState`를 가지고 있지 카드 목록을 직접적으로 가지고 있지 않습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListScreen.kt:
```diff
@@ -90,24 +84,44 @@ private fun CardListContent(
-                        baseModifier
-                    }
-                },
+                .then(if (enableScroll) Modifier.verticalScroll(scrollState) else Modifier),
```
then을 사용한 이유는 무엇인가요?

HamBeomJoon replied:
enableScroll이라는 조건에 따라 verticalScroll Modifier를 적용하려고 사용했습니다. 
사용 안하면 아래처럼 사용해야 하는건가요??
```
if (enableScroll) {
    Modifier
        .fillMaxSize()
        .padding(horizontal = 16.dp)
        .verticalScroll(scrollState)
} else {
    Modifier
        .fillMaxSize()
        .padding(horizontal = 16.dp)
}
```

moondev03 replied:
여러 가지 방법 중 `then`을 사용하신 특별한 이유가 있는지 궁금했습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListScreen.kt:
```diff
@@ -90,24 +84,44 @@ private fun CardListContent(
+}
+
+@Composable
+private fun AddCardBox(onClick: () -> Unit) {
```
`IconButton`을 사용하는 것과 어떤 차이가 있나요?

HamBeomJoon replied:
Icon: 단순히 아이콘 이미지를 표시하는 컴포넌트입니다. Modifier.clickable과 같은 추가적인 동작을 명시적으로 부여하지 않으면 클릭 이벤트에 반응하지 않습니다.

IconButton: 클릭 가능한 아이콘 버튼을 만드는 데 특화된 컴포넌트입니다. 클릭 이벤트를 처리하는 onClick 매개변수를 필수적으로 가지며, 기본적으로 **물결 효과(ripple effect)** 와 같은 시각적 피드백을 내장하고 있습니다.

+아이콘뿐만 아니라 Box 전체에 클릭이벤트를 주고싶어서 Icon을 사용하고 Box에 clickable을 달았습니다.

moondev03 replied:
미플이 정리해주신 `Icon`과 `IconButton`에 대한 내용을 살펴보면 
`Icon`은 클릭 이벤트에 반응하지 않고 오히려 `IconButton`에서 클릭 이벤트가 처리된다고 되어 있네요 🤔

> 아이콘뿐만 아니라 Box 전체에 클릭이벤트를 주고싶어서 Icon을 사용하고 Box에 clickable을 달았습니다.

잘못된 내용인 것 같아요

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListScreen.kt:
```diff
@@ -31,15 +32,17 @@ import androidx.compose.ui.tooling.preview.PreviewParameterProvider
 fun CardListScreen(
-    stateHolder: CardListStateHolder = remember { CardListStateHolder() },
     navigateToAddCard: () -> Unit,
+    stateHolder: CardListStateHolder = remember { CardListStateHolder() },
```
`remember`를 사용하면 Configuration Change가 발생할 때 UI 상태가 유지되지 않을 것 같아요.
실제로 확인해보니 유지되지 않네요 🤔

HamBeomJoon replied:
saver의 restore에서 복원하도록 수정하였습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardStateHolder.kt:
```diff
@@ -10,64 +10,99 @@ import woowacourse.payments.domain.CardNumber
     }
 
-    fun updatePassword(newPassword: Password?) {
+    fun updatePassword(newPassword: String) {
         uiState = uiState.copy(password = newPassword)
+        validate()
```
비밀번호를 입력할 때만 검증을 수행하는 이유가 무엇인가요?

HamBeomJoon replied:
아 처음에 모든 항목에 다 넣었다가 마지막에 다 뺐는데, 누락된것같네요

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardStateHolder.kt:
```diff
@@ -10,64 +10,99 @@ import woowacourse.payments.domain.CardNumber
+import woowacourse.payments.ui.util.getParcelableArrayListCompat
+import woowacourse.payments.ui.util.getParcelableExtraCompat
 
 class AddCardStateHolder(
```
미플이 생각하는 `StateHolder`의 역할은 무엇인가요?

HamBeomJoon replied:
UI의 상태를 관리하고 비즈니스 로직을 처리, 상태 보존하는 역할이라고 생각합니다. uiState -> uiModel 매핑함수는 uiState로 이동시켰습니다.

moondev03 replied:
공식 문서에서는 상태 홀더를 2가지로 구분하고 있어요.
참고하시면 좋을 것 같습니다 :)

### 비즈니스 로직 상태 홀더 (Business-logic StateHolder)
- 화면(UI)의 데이터 상태(screen-UIstate)를 생성하는 책임이 있음.
- 사용자 이벤트를 처리하고, 도메인/데이터 계층에서 데이터를 가져와서 UI 상태로 변환.  ￼
- 구성 변경(Activity 재생성 등) 시에도 상태가 유지되어야 함.  ￼
- 탐색 네비게이션(graph) 등의 변화에도 상태가 오래 살아야 할 수 있음.  ￼

### UI 로직 상태 홀더 (UI-logic StateHolder)
- UI 요소 상태(element state) 또는 UI 화면에서의 표시 방식 조정 등 UI와 직접 관계 있는 작은 로직을 가짐.
- 보통 구성 변경시 유지될 필요가 없는, UI 수명 주기에 종속적인 상태들 (e.g. Compose의 UI 요소 상태, remember / rememberSaveable 등)  ￼
- 여러 UI에서 재사용 가능할 수도 있음. 예: 필터 표시 UI나 검색창 UI 요소 등의 일부분.  ￼

<br>

> 두 가지 종류의 로직이 모두 UI 상태 생성 파이프라인에 적용되는 경우 비즈니스 로직이 항상 UI 로직보다 먼저 적용되어야 합니다. UI 로직 다음에 비즈니스 로직을 적용하려고 하면 비즈니스 로직이 UI 로직에 종속된다는 것을 의미합니다.

<p align="center">
  <img width="300" height="321" alt="Image" src="https://github.com/user-attachments/assets/e2706e93-3999-468b-b1cd-82b7ab7ec1cc" />
</p>

> UI 로직 상태 홀더가 데이터 또는 도메인 레이어의 정보에 종속되는 경우 비즈니스 로직 상태 홀더에서 UI 로직 상태 홀더로 이 정보를 전달해야 합니다. 이는 비즈니스 로직 상태 홀더가 UI 수명 주기와 무관하므로 UI 로직 상태 홀더보다 수명이 길기 때문입니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/model/BankUiModel.kt:
```diff
@@ -0,0 +1,29 @@
+import androidx.compose.ui.graphics.toArgb
+import kotlinx.parcelize.Parcelize
+
+@Parcelize
+data class BankUiModel(
+    val name: String,
+    val color: Int,
+    val logoRes: Int,
+) : Parcelable {
+    val backgroundColor: Color
+        get() = Color(color)
+
+    companion object {
+        fun create(
+            name: String,
+            color: Color,
+            logoRes: Int,
+        ): BankUiModel =
+            BankUiModel(
+                name = name,
+                color = color.toArgb(),
+                logoRes = logoRes,
+            )
+    }
+}
```
이런 구조로 `BankUiModel`를 구현한 이유가 궁금합니다!

HamBeomJoon replied:
Color 객체는 Parcelable을 지원하지 않기 때문에, Int 값으로 변환해야만 데이터를 안전하게 직렬화하여 전달할 수 있습니다.

moondev03 replied:
`data class`라 팩토리 메서드를 강제할 수 없는 점이 아쉽네요.
`TypeParceler`도 한 번 살펴보시면 도움이 될 것 같습니다.

그리고 logoRes는 Color처럼 문제될 부분이 없으니 `@DrawableRes`를 붙여주셔도 좋을 것 같아요.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/model/CardUiModel.kt:
```diff
@@ -2,27 +2,94 @@ package woowacourse.payments.ui.model
+        }
+
+    val formattedExpired: String
+        get() = expired.chunked(2).joinToString(" / ")
```
만약 포맷 요구사항이 변경된다면 어떻게 될까요?

HamBeomJoon replied:
확장함수로 분리해보긴했는데, CardUiModel 안에 함수로 놓는것과 확장함수로 놓는 것 중엔 어떤게 낫다고 생각하시나요??
8176931d84544b8544bc5813c1b1df8db0724ab6

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardError.kt:
```diff
@@ -1,8 +1,15 @@
+import kotlinx.parcelize.Parcelize
+import woowacourse.payments.R
+
+@Parcelize
+enum class AddCardError(
+    val messageRes: Int,
+) : Parcelable {
+    CARD_NUMBER_INVALID(R.string.card_number_invalid),
+    EXPIRED_INVALID(R.string.expired_invalid),
+    OWNER_INVALID(R.string.card_owner_invalid),
+    PASSWORD_INVALID(R.string.password_invalid),
```
에러 메시지의 종류가 많아진다면 어떻게 처리할 수 있을까요?

HamBeomJoon replied:
에러 메시지를 기능별로 sealed Class으로 관리할 수 있을 것 같아요. 네트워크 관련 에러는 NetworkError에, 유효성 검사 관련 에러는 InputValidationError에 넣는 식?

```
sealed class PaymentError {
    sealed class InputValidationError(val messageRes: Int) : PaymentError() {
        object CardNumberInvalid : InputValidationError(R.string.card_number_invalid)
        object ExpiredInvalid : InputValidationError(R.string.expired_invalid)
    }

    sealed class NetworkError(val messageRes: Int) : PaymentError() {
        object Timeout : NetworkError(R.string.network_timeout)
        data class HttpError(val code: Int) : NetworkError(R.string.http_error)
    }
}
```

moondev03 reviewed (changes requested):
3단계 구현 고생하셨습니다 :)

HamBeomJoon committed:
refactor: OutlinedTextField 기본 text Color 적용

HamBeomJoon committed:
fix: 은행 선택 시 BottomSheet가 사라지도록 수정

HamBeomJoon committed:
feat: BankItem에 modifier 파라미터 추가

HamBeomJoon committed:
feat: 은행 선택 아이템 description에 은행 이름 추가

HamBeomJoon committed:
refactor: 화면 회전 시 상태를 유지하도록 StateHolder 및 Saver 수정

- CardListActivity와 AddCardActivity에서 StateHolder를 rememberSaveable로 관리하도록 변경
- CardListStateHolder의 Saver가 복원 시 초기 카드 목록을 사용하도록 수정
- AddCardStateHolder의 Saver가 AddCardUiState를 직접 저장하고 복원하도록 변경
- AddCardScreen에서 StateHolder를 매개변수로 받도록 수정하고, uiState를 StateHolder로부터 가져오도록 변경

HamBeomJoon committed:
feat: AddCardUiState를 Parcelable로 변경, CardListScreen, AddCardScreen 상태 홀더 파라미터 위치 변경, string.xml에 문자열 리소스 추가

- AddCardUiState를 Parcelable로 변경하여 상태를 저장하고 복원할 수 있도록 함
- CardListScreen, AddCardScreen의 상태 홀더 파라미터 위치를 변경하여 코드 가독성을 향상
- string.xml에 카드 추가 버튼 설명을 위한 문자열 리소스 추가

HamBeomJoon committed:
feat: 카드 추가 버튼 설명 추가 및 카드 추가 상태 홀더 유효성 검사 수정

- CardListScreen의 카드 추가 버튼에 contentDescription을 추가했습니다.
- AddCardStateHolder의 유효성 검사 로직을 수정하여 유효성 검사 결과를 반환하도록 했습니다.
- AddCardScreen에서 카드 저장 시 유효성 검사 결과를 활용하도록 수정했습니다.

HamBeomJoon committed:
refactor: stateHolder 상태관리와 비즈니스 로직만 수행하도록 수정

HamBeomJoon committed:
refactor: `BankType.toPresentation` 함수 이동 및 `formattedNumber`, `formattedExpired`를 확장 함수로 변경

moondev03 commented on app/src/main/java/woowacourse/payments/domain/BankType.kt:
```diff
@@ -0,0 +1,13 @@
+    LOTTE,
+    HANA,
+    KB,
+    NOT_SELECTED,
```
저는 `NOT_SELECTED`라는 상태가 도메인이 아니라고 생각해요.
미플은 어떻게 생각하시나요?

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/BankSelectBottomSheet.kt:
```diff
@@ -0,0 +1,75 @@
+                verticalArrangement = Arrangement.spacedBy(32.dp),
+            ) {
+                BankType.entries.filterNot { it == NOT_SELECTED }.forEach { bank ->
+                    BankItem(
+                        bankType = bank,
+                        onClick = {
+                            onBankSelected(bank)
+                            coroutineScope.launch {
+                                sheetState.hide()
+                                onDismiss()
+                            }
+                        },
```
`hide()`와 `onDismiss()` 호출을 여기서 하도록 변경하신 이유가 궁금합니다.

HamBeomJoon replied:
AddScreen으로 끌어올렸습니다 👍

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardStateHolder.kt:
```diff
@@ -1,76 +1,62 @@
+        uiState = uiState.copy(bankUiModel = newBank)
+    }
+
+    fun validate(): Boolean {
+        val errors = mutableSetOf<AddCardError>()
+        val cardNumberValue = uiState.cardNumber
+        val expiredValue = uiState.expired
+        val ownerValue = uiState.cardOwner
+        val passwordValue = uiState.password
+
+        if (!CardNumber(cardNumberValue).isValid) errors.add(AddCardError.CARD_NUMBER_INVALID)
+        if (!Expired(expiredValue).isValid) errors.add(AddCardError.EXPIRED_INVALID)
+        if (!CardOwner(ownerValue).isValid) errors.add(AddCardError.OWNER_INVALID)
+        if (!Password(passwordValue).isValid) errors.add(AddCardError.PASSWORD_INVALID)
+
+        uiState = uiState.copy(errors = errors, submitted = true)
+        return errors.isEmpty()
     }
```
실행해 보니 은행을 선택하지 않아도 카드가 생성되네요 🤔

HamBeomJoon replied:
은행을 선택하지않는다는게 그냥 bottomsheet에서 위에 빈칸을 누른건가요? 저는 막혀있어서 생성안되긴합니다

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListStateHolder.kt:
```diff
@@ -7,20 +7,30 @@ import androidx.compose.runtime.setValue
+        uiState =
+            uiState.copy(
+                cards = newCards,
+                showAddButton = newCards.size > 1,
```
`showAddButton`이라는 상태의 결정을 UiStete에서 처리되도록 할 수 있을 것 같아요

moondev03 commented on app/src/main/java/woowacourse/payments/ui/theme/Color.kt:
```diff
@@ -9,3 +9,14 @@ val Pink80 = Color(0xFFEFB8C8)
 val PurpleGrey40 = Color(0xFF625b71)
 val Pink40 = Color(0xFF7D5260)
+
+val GOLD = Color(0xFFCBBA64)
+val GRAY = Color(0xFFC8C8C8)
+val BC = Color(0xFFF04651)
+val SHINHAN = Color(0xFF293A94)
+val KAKAO = Color(0xFFFAE100)
+val HYUNDAI = Color(0xFF000000)
+val WOORI = Color(0xFF2371B3)
+val LOTTE = Color(0xFFE21E26)
+val HANA = Color(0xFF00908F)
+val KB = Color(0xFF695F54)
```
미플은 언제 색상 리소스를 분리하시나요?

moondev03 reviewed (approved):
미플 리뷰 반영 고생하셨습니다 👍 
리뷰 확인해주시고 다음 단계에서 반영 부탁드려요:)
